### PR TITLE
HGCAL trigger sim clustering

### PR DESCRIPTION
**Pull request by jbsauvan**
Implements the production of sim clusters based on HGCAL trigger cells

### CODE REVIEW

File: DataFormats/L1THGCal/src/HGCalCluster.cc [hunk 1]
```diff
@@ -31,3 +31,4 @@ bool HGCalCluster::operator<(const HGCalCluster& cl) const
   }
   return res;
 }
+
```
Comment (kpedro88) [inline DataFormats/L1THGCal/src/HGCalCluster.cc]: please avoid whitespace-only changes

File: L1Trigger/L1THGCal/interface/fe_codecs/HGCalBestChoiceCodecImpl.h [hunk 2]
```diff
@@ -20,6 +20,7 @@ struct HGCalBestChoiceDataPayload
     { 
         payload.fill(0);
     }
+
```
Comment (kpedro88) [inline L1Trigger/L1THGCal/interface/fe_codecs/HGCalBestChoiceCodecImpl.h]: please avoid whitespace-only changes

File: L1Trigger/L1THGCal/plugins/be_algorithms/HGCalSimCluster.cc [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+
+}// namespace
+
+
+// define plugins, template needs to be spelled out here, in order to allow the compiler to compile, and the factory to be populated
+//
+typedef HGCalTriggerBackend::HGCalTriggerSimCluster<HGCalTriggerCellBestChoiceCodec,HGCalTriggerCellBestChoiceDataPayload> HGCalTriggerSimClusterBestChoice;
+DEFINE_EDM_PLUGIN(HGCalTriggerBackendAlgorithmFactory, HGCalTriggerSimClusterBestChoice,"HGCalTriggerSimClusterBestChoice");
+
+// Local Variables:
```
Comment (kpedro88) [inline L1Trigger/L1THGCal/plugins/be_algorithms/HGCalSimCluster.cc]: I highly doubt these comments need to be tracked in official-cmssw

File: DataFormats/L1THGCal/interface/ClusterShapes.h [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        void operator+=(const ClusterShapes &); 
+        ClusterShapes& operator=(const ClusterShapes &) = default;
+    };
+
+}; // end namespace
+
+#endif
+
+// -----------------------------------
+// Local Variables:
```
Comment (kpedro88) [inline DataFormats/L1THGCal/interface/ClusterShapes.h]: I highly doubt these comments need to be tracked in official-cmssw

File: L1Trigger/L1THGCal/test/testTriggerSimCluster.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+c2.cd(1)
+hN.Draw("HIST")
+c2.cd(2)
+hSEE.Draw("HIST")
+c2.cd(3)
+hSPP.Draw("HIST")
+c2.cd(4)
+hSRR.Draw("HIST")
+
+#c3=ROOT.TCanvas("c3","c3")
```
Comment (kpedro88) [inline L1Trigger/L1THGCal/test/testTriggerSimCluster.py]: In general, commented-out code should be removed

File: L1Trigger/L1THGCal/plugins/be_algorithms/HGCalSimCluster.cc [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+                          continue;
+                        }
+                        unsigned recoCell = HGCalDetId(mysubdet,zp,layer,subsec,sec,cell);
+                        hgc_simhit_energy[recoCell] += simhit.energy();
+                    }
+                }
+
+                if (bh_simhits_h.isValid() ) /// FIXME TODO
+                {
+                       throw cms::Exception("Not Implemented")<<"HGCalTriggerSimCluster: BH simhits not implemnted";
```
Comment (kpedro88) [inline L1Trigger/L1THGCal/plugins/be_algorithms/HGCalSimCluster.cc]: Out of curiosity, when is the BH implementation expected?

**Comment by jbsauvan** [comment]
Hi @kpedro88, thanks for the comments!
I have made a bit of cleaning.

Concerning the BH implementation, this is difficult to answer because absolutely no study has been done on the BH in the trigger so far.
As a first step we could have a trivial trigger path that just copy the cells into "trigger cells". But even that  will need a bit of thinking / study concerning:
1. The definition of "modules" in that case (the trigger cell creation and selection is done within module regions defined in the trigger geometry class, but it is not clear how BH modules will be defined)
2. The coding of the energy, in order to make sure we have something compatible with the EE and FH parts

**Comment by kpedro88** [comment]
@jbsauvan you can close this one, it will just go in 91X

**Pull request closed without merging**